crates/polars-core/src/series/arithmetic/list_borrowed.rs

```diff
@@ -1,177 +1,1026 @@
+    {
+        #[inline(never)]
+        fn check_mismatch_pos(
+            mismatch_pos: usize,
+            offsets_lhs: &OffsetsBuffer<i64>,
+            offsets_rhs: &OffsetsBuffer<i64>,
+        ) -> PolarsResult<()> {
+            if mismatch_pos < offsets_lhs.len_proxy() {
+                // RHS could be broadcasted
+                let len_r = offsets_rhs.length_at(if offsets_rhs.len_proxy() == 1 {
+                    0
+                } else {
+                    mismatch_pos
+                });
+                polars_bail!(
+                    ShapeMismatch:
+                    "list lengths differed at index {}: {} != {}",
+                    mismatch_pos,
+                    offsets_lhs.length_at(mismatch_pos), len_r
+                )
+            }
+            Ok(())
+        }
+
+        let mut arr_lhs = {
+            let ca: &ChunkedArray<T> = prim_s_lhs.as_ref().as_ref();
+            assert_eq!(ca.chunks().len(), 1);
+            ca.downcast_get(0).unwrap().clone()
+        };
+
+        let mut arr_rhs = {
+            let ca: &ChunkedArray<T> = prim_s_rhs.as_ref().as_ref();
+            assert_eq!(ca.chunks().len(), 1);
+            ca.downcast_get(0).unwrap().clone()
+        };
+
+        match (&self.op_apply_type, &self.broadcast) {
+            // We skip for this because it dispatches to `ArithmeticKernel`, which handles the
+            // validities for us.
+            (BinaryOpApplyType::ListToPrimitive, Broadcast::Right) => {},
+            _ if self.list_to_prim_lhs.is_none() => self
+                .op
+                .prepare_numeric_op_side_validities::<T>(&mut arr_lhs, &mut arr_rhs, self.swapped),
+            (BinaryOpApplyType::ListToPrimitive, Broadcast::NoBroadcast) => {},
+            _ => unreachable!(),
+        }
+
+        //
+        // General notes
+        // * Lists can be:
+        //   * Sliced, in which case the primitive/leaf array needs to be indexed starting from an
+        //     offset instead of 0.
+        //   * Masked, in which case the masked rows are permitted to have non-matching widths.
+        //
+
+        let out = match (&self.op_apply_type, &self.broadcast) {
+            (BinaryOpApplyType::ListToList, Broadcast::NoBroadcast) => {
+                let offsets_lhs = &self.data_lhs.0[0];
+                let offsets_rhs = &self.data_rhs.0[0];
+
+                assert_eq!(offsets_lhs.len_proxy(), offsets_rhs.len_proxy());
+
+                // Output primitive (and optional validity) are aligned to the LHS input.
+                let n_values = arr_lhs.len();
+                let mut out_vec: Vec<T::Native> = Vec::with_capacity(n_values);
+                let out_ptr: *mut T::Native = out_vec.as_mut_ptr();
+
+                // Counter that stops being incremented at the first row position with mismatching
+                // list lengths.
+                let mut mismatch_pos = 0;
+
+                with_match_numeric_list_op!(&self.op, self.swapped, |$OP| {
```

Comment:
I kinda feel like this should not be here. Instead, this should be in `polars-compute`.

Reply:
I tried to move it, but I ran into some difficulty as `polars-compute` doesn't have access to the `Series` struct - I think we can leave it here for now?